### ISSUE TITLE
retrieve html

### ISSUE THREAD

**Issue by ilrisa** [cwollenhaupt/fpCefSharp#10]
Good morning.
Thanks to your great work I was able to upload html and PDF files to my Fox windows.

Is there a method to be able to retrieve the HTML code of the page loaded in the ole control?

Thanks

**Comment by dsdsdfsdfsfsdfsfsdf** [comment]
HTML code should work via javascript. I tried it but it doesn't work for me. OLE object has CHROMIUM object, but Document object he is not here. Does anyone have another solution?

**Comment by cwollenhaupt** [comment]
As you said, you have to execute java script in order to access the DOM including the HTML text (either innerHTML or outerHTML). Here's a working example (working for me, that is 😉):

```
*========================================================================================
* Retrieve HTML from a loaded document
*========================================================================================
	
	Public goForm
	Set Procedure To CefSharpBrowser.fxp ADDITIVE
	goForm = CreateObject ("DemoForm")
	goForm.Show ()
	

Define Class DemoForm as Form
	
	Height = 440
	Width = 860

	Add Object oleBrowser as OleControl With ;
		 OleClass = "fpDotNet.DotNetContainer" ;
		,Top = 5 ;
		,Left = 5 ;
		,Height = 395 ;
		,Width = 850 ;
		,Anchor = 15
		
	Add Object cefSharpBrowser as CefSharpBrowser 
	
	Add Object cmdGetHtml as CommandButton with ;
		 Caption = "Get HTML" ;
		,Top = 405 ;
		,Left = 5 ;
		,Height = 25 ;
		,Anchor = 6

	index_html = ""
	
Procedure Init
	TEXT TO This.index_html NoShow
		<html>
		<form id="myform">
		  Text 1: <input type="text" id="text1" name="text1" value="Hello"><br>
		  Text 2: <input type="text" id="text2" name="text2" value="World"><br>
		  <input type="submit" value="Submit">
		</form>
		</html>	
	ENDTEXT 

	This.CefSharpBrowser.BindtoHost ( ;
		 This.oleBrowser ;
		,"https://myApp/index.html" ;
		,This ;
	)
	
Procedure cmdGetHtml.Click
	Local lcResult
	lcResult = Thisform.EvaluateScript ( ;
		"document.getElementById('myform').innerHTML" ;
	)
	MessageBox (m.lcResult)	

Procedure EvaluateScript (tcScript)
	
	Local loBridge, loBrowser, loFrame
	loBridge = This.CefSharpBrowser.DotNet ()
	loBrowser = loBridge.InvokeMethod (This.CefSharpBrowser.oChromium, "GetBrowser")
	loFrame = loBridge.GetProperty (m.loBrowser, "MainFrame")
	
	Local loTask, loResponse, luResult
	loTask = loBridge.InvokeMethod (m.loFrame, "EvaluateScriptAsync" ;
		,m.tcScript ;
		,"about:blank" ;
		,1 ;
		,null ;
		,.F. ;
	)
	loResponse = loBridge.GetProperty (m.loTask, "Result")
	luResult = loBridge.GetProperty (m.loResponse, "Result")
	
Return m.luResult
	
Procedure QueryUnload
	This.ReleaseRefrences ()
	
Procedure Release
	This.ReleaseRefrences ()

Procedure ReleaseRefrences
	This.RemoveObject ("oleBrowser")
	This.RemoveObject ("CefSharpBrowser")
		
EndDefine 
```

This code requires the cefSharpBrowser.prg to be available and the fpDotNet.DotNetContainer. If you can run any of the other demos then you should be all set.

Click on the button at the bottom, and a messagebox displays the HTML code for the form. This corresponds to the following javascript expression:

`document.getElementById('myform').innerHTML`

You can replace this with any other expression including calling any of the javascript functions that are defined in the page. Please note that the return value cannot be an object. It must be a number, a string, or any other basic data type that is supported.

Most of the code is just to setup the browser control. The relevant code to access javascript and with that access anything in the DOM is in the EvaluateScript method that is called from the button's click event.

**Comment by dsdsdfsdfsfsdfsfsdf** [comment]
> 
This example works! Thank you for everyone. In the EvaluateScript (tcScript) procedure, there is a call InvokeMethod () method, but it is actually a call to InvokeMethod_FiveParms () method, so my Script didn't work for me either. When calling e.g. InvokeMethod_FiveParms (m.loFrame, "EvaluateScriptAsync", [document.all ("f"). Submit ()], "about: blank", 1, NULL, .F.) I did not find the purpose of the parameter "about : blank "and another behind it with a value of 1. I tried to use them empty as well and they work correctly as well. Thanks for the early resolution.
